DevEdu.API/DevEdu.Business/Services/TagService.cs

```diff
@@ -7,18 +8,25 @@ namespace DevEdu.Business.Services
     public class TagService : ITagService
     {
         private readonly ITagRepository _repository;
+        private readonly ITagValidationHelper _tagValidationHelper;
 
-        public TagService(ITagRepository repository)
+        public TagService(ITagRepository repository, ITagValidationHelper tagValidationHelper)
         {
             _repository = repository;
+            _tagValidationHelper = tagValidationHelper;
         }
 
         public int AddTag(TagDto dto) => _repository.AddTag(dto);
 
-        public void DeleteTag(int id) => _repository.DeleteTag(id);
+        public void DeleteTag(int id)
+        {
+            _tagValidationHelper.CheckTagExistence(id);
+            _repository.DeleteTag(id);
+        }
 
         public TagDto UpdateTag(TagDto dto, int id)
         {
+            _tagValidationHelper.CheckTagExistence(id);
             dto.Id = id;
             _repository.UpdateTag(dto);
             return GetTagById(id);

```

Comment:
you'd better use repository's method here
and service's GetTagById must return EntityNotFoundException if nothing was found by id